sdks/java/extensions/smb/src/main/java/org/apache/beam/sdk/extensions/smb/SMBFilenamePolicy.java

```diff
@@ -107,20 +116,47 @@ ResourceId forBucket(BucketShardId id, BucketMetadata<?, ?> metadata) {
               : String.format(NUMERIC_BUCKET_TEMPLATE, id.getBucketId(), metadata.getNumBuckets());
 
       final String timestamp = doTimestampFiles ? Instant.now().toString(TEMPFILE_TIMESTAMP) : "";
-      String filename =
+      final String filename =
           String.format(
               BUCKET_SHARD_TEMPLATE,
               bucketName,
               id.getShardId(),
               metadata.getNumShards(),
               filenameSuffix);
-      return filenamePrefix.resolve(timestamp + filename, StandardResolveOptions.RESOLVE_FILE);
+
+      final ResourceId file =
+          filenamePrefix.resolve(timestamp + filename, StandardResolveOptions.RESOLVE_FILE);
+
+      return new ReadableFile(getMetadata(file), compression);
+    }
+
+    ReadableFile forMetadata() {
+      final String timestamp = doTimestampFiles ? Instant.now().toString(TEMPFILE_TIMESTAMP) : "";
+      final ResourceId file =
+          filenamePrefix.resolve(
+              timestamp + METADATA_FILENAME, StandardResolveOptions.RESOLVE_FILE);
+
+      return new ReadableFile(getMetadata(file), Compression.UNCOMPRESSED);
     }
 
-    ResourceId forMetadata() {
-      String timestamp = doTimestampFiles ? Instant.now().toString(TEMPFILE_TIMESTAMP) : "";
-      return filenamePrefix.resolve(
-          timestamp + METADATA_FILENAME, StandardResolveOptions.RESOLVE_FILE);
+    static Metadata getMetadata(ResourceId resourceId) {
+      try {
+        final List<Metadata> metadata =
+            FileSystems.match(resourceId.toString(), EmptyMatchTreatment.ALLOW).metadata();
+
+        if (metadata.isEmpty()) {
```

Comment:
This branch is for writes only right? It'll be nice to avoid the extra `FileSystems.match` since we're only dealing with new files for writes.